test/e2etest/kubetest/results_evaluator.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/gardener/test-infra/test/e2etest/util"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	E2eLogFileNamePattern   = "e2e.log$"
+	JunitXmlFileNamePattern = `junit_\d+.xml$`
+	TestSummaryFileName     = "test_summary.json"
+	MergedJunitXmlFile      = "junit_01.xml"
+	MergedE2eLogFile        = "e2e.log"
+)
+
+var mergedJunitXmlFilePath = filepath.Join(config.ExportPath, MergedJunitXmlFile)
+var MergedE2eLogFilePath = filepath.Join(config.ExportPath, MergedE2eLogFile)
+
+// Analyze analyzes junit.xml files and e2e.log files, which are dumped by kubetest and provides a resulting test suite summary and results for each testcase individually. These results are then written to the export dir as files.
+func Analyze(kubetestResultsPath string) Summary {
+	log.Info("Analyze e2e.log and junit.xml files")
+	e2eLogFilePaths := util.GetFilesByPattern(kubetestResultsPath, E2eLogFileNamePattern)
+	summary := analyzeE2eLogs(e2eLogFilePaths)
+	junitXMLFilePaths := util.GetFilesByPattern(kubetestResultsPath, JunitXmlFileNamePattern)
+	analyzeJunitXMLs(junitXMLFilePaths, summary.TestsuiteDuration)
+	log.Infof("Check out result files in %s", kubetestResultsPath)
+	return summary
+}
+
+func analyzeJunitXMLs(junitXMLFilePaths []string, durationSec int) {
+	var mergedJunitXmlResult = &JunitXMLResult{FailedTests: 0, ExecutedTests: 0, DurationFloat: 0, SuccessfulTests: 0, DurationInt: durationSec}
+	testcaseNameToTestcase := make(map[string]TestcaseResult)
+	for _, junitXMLPath := range junitXMLFilePaths {
+		file, err := os.Open(junitXMLPath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer file.Close()
+		junitXml, err := UnmarshalJunitXMLResult(file.Name())
+		if err != nil {
+			log.Fatal(errors.Wrapf(err, "Couldn't unmarshal %s", file.Name()))
+		}
+		mergedJunitXmlResult.FailedTests += junitXml.FailedTests
+		mergedJunitXmlResult.ExecutedTests += junitXml.ExecutedTests
+		mergedJunitXmlResult.SuccessfulTests += junitXml.SuccessfulTests
+		for _, testcase := range junitXml.Testcases {
+			if testcase.Skipped {
+				if _, ok := testcaseNameToTestcase[testcase.Name]; !ok {
+					testcaseNameToTestcase[testcase.Name] = testcase
+				}
+				continue
+			}
+			testcaseNameToTestcase[testcase.Name] = testcase
+			testcaseJSON, err := json.MarshalIndent(testcase, "", " ")
+			if err != nil {
+				log.Fatal(errors.Wrapf(err, "Couldn't marshal testsuite summary %s", testcaseJSON))
+			}
+
+			jsonFileName := fmt.Sprintf("test-%s.json", strconv.FormatInt(time.Now().UnixNano(), 10))
+			testcaseJsonFilePath := path.Join(config.ExportPath, jsonFileName)
+			if err := ioutil.WriteFile(testcaseJsonFilePath, testcaseJSON, 0644); err != nil {
+				log.Fatal(errors.Wrapf(err, "Couldn't write %s to file", testcaseJsonFilePath))
+			}
+		}
+	}
+	for _, testcase := range testcaseNameToTestcase {
```

Comment:
why do we need to iterate again instead of appending the directly in the loop before.
I see that this is a set and only the last result with this testcase is then appended. 
But is it possible that there are multiple results for the same name?
Do we want then the last result to win?

Reply:
Running e2etest parallely creates #Go-Routines junit_%n%.xml files. Each junit.xml files contains the testcases that have been executed and additionally all remaining testcases with XML element <skipped>.  Hence, a testcase can be skipped in one junit.xml file, but executed in another (the executed one we want to keep in the merged junit.xml file). Only the testcases, that are skipped in every junit.xml file, we write as skipped in the merged junit.xml file. If this is still unclear, let's discuss in person :-)

Reply:
ok I understand